### PR TITLE
NN-3920: Add offence details support to drafts

### PR DESCRIPTION
**Pull request by sp-watson**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt [hunk 1]
```diff
@@ -65,6 +67,28 @@ class DraftAdjudicationService(
...
+    val draftAdjudication = draftAdjudicationRepository.findById(id).orElseThrow { throwEntityNotFoundException(id) }
+
+    val newValuesToStore = offenceDetails.map {
+      Offence(
+        offenceCode = it.offenceCode,
+        victimPrisonersNumber = it.victimPrisonersNumber,
+      )
+    }.toMutableList()
+    if (draftAdjudication.offenceDetails != null) {
+      draftAdjudication.offenceDetails!!.removeAll(draftAdjudication.offenceDetails!!)
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt:83]: This is weird - but it was the only 1-liner I could find to do this.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/DraftAdjudicationControllerTest.kt [hunk 2]
```diff
@@ -162,8 +163,6 @@ class DraftAdjudicationControllerTest : TestControllerBase() {
         .andExpect(jsonPath("$.draftAdjudication.incidentDetails.dateTimeOfIncident").value("2010-10-12T10:00:00"))
         .andExpect(jsonPath("$.draftAdjudication.incidentDetails.handoverDeadline").value("2010-10-14T10:00:00"))
         .andExpect(jsonPath("$.draftAdjudication.incidentDetails.locationId").value(1))
-        .andExpect(jsonPath("$.draftAdjudication.incidentRole.roleCode").value("25a"))
```
Comment (sp-watson) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/DraftAdjudicationControllerTest.kt:165]: Getting rid of checks for props - they are checked by the int tests anyway

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt [hunk 3]
```diff
@@ -100,6 +101,41 @@ class DraftAdjudicationIntTest : IntegrationTestBase() {
       .jsonPath("$.draftAdjudication.incidentDetails.locationId").isEqualTo(testAdjudication.locationId)
       .jsonPath("$.draftAdjudication.incidentRole.roleCode").isEqualTo(testAdjudication.incidentRoleCode)
       .jsonPath("$.draftAdjudication.incidentRole.associatedPrisonersNumber").isEqualTo(testAdjudication.incidentRoleAssociatedPrisonersNumber)
+      // Not implemented yet
```
Comment (sp-watson) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt:104]: This won't work until we add the offence info to the reported adjs (next PR)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt [hunk 4]
```diff
@@ -100,6 +101,41 @@ class DraftAdjudicationIntTest : IntegrationTestBase() {
...
+      .expectStatus().isCreated
+      .expectBody()
+      .jsonPath("$.draftAdjudication.id").isNumber
+      .jsonPath("$.draftAdjudication.prisonerNumber").isEqualTo(testAdjudication.prisonerNumber)
+      .jsonPath("$.draftAdjudication.incidentDetails.dateTimeOfIncident").isEqualTo(testAdjudication.dateTimeOfIncidentISOString)
+      .jsonPath("$.draftAdjudication.incidentDetails.handoverDeadline").isEqualTo(testAdjudication.handoverDeadlineISOString)
+      .jsonPath("$.draftAdjudication.incidentDetails.locationId").isEqualTo(testAdjudication.locationId)
+      .jsonPath("$.draftAdjudication.incidentRole.roleCode").isEqualTo(testAdjudication.incidentRoleCode)
+      .jsonPath("$.draftAdjudication.incidentRole.associatedPrisonersNumber").isEqualTo(testAdjudication.incidentRoleAssociatedPrisonersNumber)
+      .jsonPath("$.draftAdjudication.offenceDetails[0].offenceCode").isEqualTo(testAdjudication.offences[0].offenceCode)
```
Comment (sp-watson) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt:135]: I tried to make an abstraction for these offence checks, but i couldn't find a way of making a method that could be used in place of `jsonPath`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/entities/DraftAdjudication.kt [hunk 5]
```diff
@@ -21,6 +22,9 @@ data class DraftAdjudication(
   @OneToOne(optional = true, cascade = [CascadeType.ALL], fetch = FetchType.EAGER)
   @JoinColumn(name = "incident_role_id")
   var incidentRole: IncidentRole,
+  @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.EAGER)
+  @JoinColumn(name = "draft_adjudication_fk_id")
+  var offenceDetails: MutableList<Offence>? = null,
```
Comment (richardpopple) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/entities/DraftAdjudication.kt:27]: Not sure that its any better but I guess you could have:
var offenceDetails: MutableList<Offence> = mutableListOf<>()
Then the list wouldn't ever be nullable.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt (hunk 1 above)
Comment (richardpopple) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt:83]: So I might not be understanding this - but are we trying to empty the list out? Can 
draftAdjudication.offenceDetails!!.removeAll(draftAdjudication.offenceDetails!!)
not be replaced by
draftAdjudication.offenceDetails!!.clear()

File: src/main/resources/db/migration/V15__add_offence_table.sql [hunk 6]
```diff
@@ -0,0 +1,17 @@
+drop table if exists offence;
+
+create table offence
+(
+    id                          serial primary key,
+    draft_adjudication_fk_id    bigint references draft_adjudications (id),
+    offence_code                integer       not null,
```
Comment (richardpopple) [inline src/main/resources/db/migration/V15__add_offence_table.sql:7]: Out of interest do we have list of offence codes - just wondering if there should be a reference table for them, or if we are just leaving it free form

**Review by richardpopple — APPROVED** [review]

File: src/main/resources/db/migration/V15__add_offence_table.sql (hunk 6 above)
Comment (sp-watson) [reply src/main/resources/db/migration/V15__add_offence_table.sql:7]: I was thinking it's free form in the API - the UI will contain that lookup. It means we don't have to update the API if/when offence codes are added.

My initial thinking was that the API should contain this information, but it's the maintenance that's the problem. If you have an enum then the UI has to have a version of that enum, and map that to the UI values.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/entities/DraftAdjudication.kt (hunk 5 above)
Comment (sp-watson) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/entities/DraftAdjudication.kt:27]: Null is useful - it shows it hasn't been added yet

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt (hunk 1 above)
Comment (sp-watson) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt:83]: Argh yes.. that does seem obvious. I am not sure, but I might have tried that and had problems... I'll try again for the next PR, which is imminent.

Good call!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt (hunk 4 above)
Comment (sp-watson) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt:135]: Sure

**Pull request merged**